### PR TITLE
Adding Upgrade test

### PR DESCRIPTION
**Pull request by nrevanna**
Building this test over the functions added in PWX-32648, hence basing it on that branch.

    This test case verifies that btrfs  datastores stays on upgrade on a
    testbed that qualifies for dmthin.
    TestStorageClusterDMthinUpgrade upgrades PX to 3.1 from 3.0 and 2.13
    over two iterations. It'll also verify that btrfs was installed.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:
https://jenkins.pwx.dev.purestorage.com/job/3.1.0-dev-jobs/job/vSphere-upgrade-dmthin-opint/
```
01:15:48.626 --- PASS: TestStorageClusterDMthinUpgrade (2907.73s)
01:15:48.626     --- PASS: TestStorageClusterDMthinUpgrade/BasicUpgradeStorageClusterDmthinWithAllComponents (2907.73s)
01:15:48.629 PASS
01:15:48.629 ==================================================
01:15:48.629 --- PASS:
01:15:48.629     BasicUpgradeStorageClusterDmthinWithAllComponents [C50241]
01:15:48.629 --- SKIP:
01:15:48.629 --- FAIL:
01:15:48.629 ==================================================
01:15:48.638 
01:15:48.638 DONE 2 tests in 2908.025s
```

### CODE REVIEW

File: test/integration_test/utils/utils.go [hunk 1]
```diff
@@ -131,5 +151,20 @@ func RunPxCmdInPod(pod *v1.Pod, input ...string) (string, string, error) {
...
+	output := PxctlMetadata{}
+	output.PxStoreV2NodeCount = uint(getPxStoreV2NodeCount(t, px_status))
+	return output
+}
+
+func getPxStoreV2NodeCount(t *testing.T, px_status string) int {
+	out := strings.Split(px_status, "SchedulerNodeName")
+	require.Greater(t, len(out), 1, "Could not find SchedulerNodeName in the pxclt status output")
+	out = strings.Split(out[1], "Global Storage Pool")
+	require.Greater(t, len(out), 1, "Could not find \"Global Storage Pool\" string the pxclt status output")
```
Comment (pp511) [inline test/integration_test/utils/utils.go:167]: nit.. pxctl

File: test/integration_test/utils/utils.go [hunk 2]
```diff
@@ -131,5 +151,20 @@ func RunPxCmdInPod(pod *v1.Pod, input ...string) (string, string, error) {
...
 }
+func AnalyzePxctlStatus(t *testing.T, px_status string) PxctlMetadata {
+	output := PxctlMetadata{}
+	output.PxStoreV2NodeCount = uint(getPxStoreV2NodeCount(t, px_status))
+	return output
+}
+
+func getPxStoreV2NodeCount(t *testing.T, px_status string) int {
+	out := strings.Split(px_status, "SchedulerNodeName")
+	require.Greater(t, len(out), 1, "Could not find SchedulerNodeName in the pxclt status output")
```
Comment (pp511) [inline test/integration_test/utils/utils.go:165]: nit..pxctl

File: test/integration_test/utils/utils.go [hunk 3]
```diff
@@ -94,6 +100,20 @@ func mergeEnvVars(origList, newList []v1.EnvVar) []v1.EnvVar {
 	return mergedList
 }
 
+func RunPxCmdRetry(command ...string) (string, string, error) {
+	var out string
+	var stderr string
+	var err error
+	for i := 0; i < 4; i++ {
+		out, stderr, err = RunPxCmd(command...)
+		// Occassionally, commands are terminated due to bad connections. Let's retry them.
```
Comment (pp511) [inline test/integration_test/utils/utils.go:109]: nit.. Occasionally

**Review by pp511 — COMMENTED** [review]
I see a lot of overlap with the other PR. PWX-32648: Adding Install/Uninstall test. Would be easier to review once rebased.  In the meanwhile, just some nits..

**Comment by nrevanna** [comment]
@pp511 it should look better now.
I had taken care of it by rebasing against a feature branch. Looks like something got messed up. Now it should be cleaner.

File: test/integration_test/basic_dmthin_test.go [hunk 4]
```diff
@@ -131,6 +134,38 @@ func generateTestCases(opt OptionsArr, idx uint) map[string]types.TestCase {
...
+			return cluster
+		},
+		ShouldSkip: func(tc *types.TestCase) bool {
+			if len(ci_utils.PxUpgradeHopsURLList) == 0 {
+				logrus.Info("--px-upgrade-hops-url-list is empty, cannot run BasicUpgradeStorageClusterDmthinWithAllComponents test")
+				return true
+			}
+			k8sVersion, _ := k8sutil.GetVersion()
+			pxVersion := ci_utils.GetPxVersionFromSpecGenURL(ci_utils.PxUpgradeHopsURLList[0])
+			return k8sVersion.GreaterThanOrEqual(k8sutil.K8sVer1_22) && pxVersion.LessThan(pxVer2_9)
```
Comment (pp511) [inline test/integration_test/basic_dmthin_test.go:163]: Nit... DMthin is supported from 3.0.0 onwards. Let's update the pxVersion check.

File: test/integration_test/basic_dmthin_test.go [hunk 5]
```diff
@@ -290,3 +331,46 @@ func BasicInstallDmthin(tc *types.TestCase) func(*testing.T) {
 		UninstallPX(t, cluster)
 	}
 }
+
+// PxUpgradeHopsURLList is interpreted differently in this test as compared to the original intent
+// Each member of the hop list serves as the base PX version to upgrade from to the latest release version.
+func BasicUpgradeStorageClusterDmthin(tc *types.TestCase) func(*testing.T) {
```
Comment (pp511) [inline test/integration_test/basic_dmthin_test.go:337]: nit.. This test is not using DMThin in the base image after upgrade. It is checking for a negative scenario where a cluster doesn't get upgraded to dmthin.  Either provide a comment or rename the function appropriately.

**Review by pp511 — APPROVED** [review]
Just two nits. Looks good.

**Pull request merged**